### PR TITLE
Flax T5

### PR DESCRIPTION
**Pull request by thevasudevgupta**
# What does this PR do?

This PR will add T5 in Flax/Jax.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests), Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed.

@patrickvonplaten

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -233,7 +234,7 @@ def run(self):
     extras["flax"] = []  # jax is not supported on windows
 else:
     extras["retrieval"] = deps_list("faiss-cpu", "datasets")
-    extras["flax"] = deps_list("jax", "jaxlib", "flax")
+    extras["flax"] = deps_list("jax", "jaxlib", "flax", "optax")
```
Comment (patrickvonplaten) [inline setup.py:237]: optax is required for any flax training and is very lightweight. I think it is ok to put it here, no?

File: src/transformers/models/auto/modeling_flax_auto.py [hunk 2]
```diff
@@ -110,11 +114,18 @@
         (RobertaConfig, FlaxRobertaForMaskedLM),
         (BertConfig, FlaxBertForMaskedLM),
         (BigBirdConfig, FlaxBigBirdForMaskedLM),
-        (BartConfig, FlaxBartForConditionalGeneration),
```
Comment (patrickvonplaten) [inline src/transformers/models/auto/modeling_flax_auto.py:113]: @patil-suraj, I don't think we should have Bart in the masked-lm auto class even if this is the case for PyTorch. It should be in pretraining and seq2seq LM IMO.

File: src/transformers/models/bart/modeling_flax_bart.py [hunk 3]
```diff
@@ -229,7 +229,6 @@ class FlaxBartAttention(nn.Module):
     embed_dim: int
     num_heads: int
     dropout: float = 0.0
-    is_decoder: bool = False
```
Comment (patrickvonplaten) [inline src/transformers/models/bart/modeling_flax_bart.py:232]: This is actually not needed

File: tests/test_modeling_flax_bart.py [hunk 4]
```diff
@@ -72,7 +72,7 @@ def prepare_bart_inputs_dict(
     }
 
 
-class FlaxBartModelTester(unittest.TestCase):
+class FlaxBartModelTester:
```
Comment (patrickvonplaten) [inline tests/test_modeling_flax_bart.py:75]: This class doesn't have to be a `unittest.TestCase` class

File: setup.py (hunk 1 above)
Comment (patrickvonplaten) [reply setup.py:237]: @LysandreJik @sgugger @patil-suraj

File: tests/test_modeling_flax_t5.py [hunk 5]
```diff
@@ -0,0 +1,513 @@
...
+    def test_small_generation(self):
+        model = FlaxT5ForConditionalGeneration.from_pretrained("t5-small")
+        model.config.max_length = 8
+        model.config.num_beams = 1
+        model.config.do_sample = False
+        tokenizer = T5Tokenizer.from_pretrained("t5-small")
+
+        input_ids = tokenizer("summarize: Hello there", return_tensors="np").input_ids
+
+        sequences = model.generate(input_ids, trace=False).sequences
```
Comment (patrickvonplaten) [inline tests/test_modeling_flax_t5.py:469]:
```suggestion
        sequences = model.generate(input_ids, trace).sequences
```

File: tests/test_modeling_flax_t5.py [hunk 6]
```diff
@@ -0,0 +1,513 @@
...
+        loss = optax.softmax_cross_entropy(logits, onehot(labels, logits.shape[-1])).mean()
+
+        mtf_score = -(labels.shape[-1] * loss.item())
+
+        EXPECTED_SCORE = -60.7397
+        self.assertTrue(abs(mtf_score - EXPECTED_SCORE) < 1e-4)
+
+    @slow
+    def test_small_generation(self):
+        model = FlaxT5ForConditionalGeneration.from_pretrained("t5-small")
```
Comment (patrickvonplaten) [inline tests/test_modeling_flax_t5.py:461]: this test makes sure that Flax == PyTorch

File: tests/test_modeling_flax_t5.py [hunk 7]
```diff
@@ -0,0 +1,513 @@
...
+        input_ids = tokenizer("summarize: Hello there", return_tensors="np").input_ids
+
+        sequences = model.generate(input_ids, trace=False).sequences
+
+        output_str = tokenizer.batch_decode(sequences, skip_special_tokens=True)[0]
+        self.assertTrue(output_str == "Hello there!")
+
+    @slow
+    def test_summarization(self):
+        model = FlaxT5ForConditionalGeneration.from_pretrained("t5-base")
```
Comment (patrickvonplaten) [inline tests/test_modeling_flax_t5.py:476]: Agressive summarizaton test

**Comment by patrickvonplaten** [comment]
Exceptionally merging already given the time constraint of the Flax/JAX community week announcement. 
@patil-suraj @sgugger, I would be very happy if you could nevertheless take a look after merge so that I can correct suggestions in a follow-up PR before the next transformers release.

File: src/transformers/models/t5/modeling_flax_t5.py [hunk 8]
```diff
@@ -0,0 +1,1584 @@
...
+        output_hidden_states=None,
+        return_dict=None,
+        deterministic: bool = True,
+    ):
+        r"""
+        Returns:
+
+        Example::
+
+            >>> from transformers import T5Tokenizer, T5Model
```
Comment (patil-suraj) [inline src/transformers/models/t5/modeling_flax_t5.py:1245]: T5Model => FlaxT5Model

File: src/transformers/models/t5/modeling_flax_t5.py [hunk 9]
```diff
@@ -0,0 +1,1584 @@
...
+            attention_mask,
+            decoder_input_ids,
+            decoder_attention_mask,
+        )["params"]
+
+    def __call__(
+        self,
+        input_ids: jnp.ndarray,
+        attention_mask: Optional[jnp.ndarray] = None,
+        decoder_input_ids: Optional[jnp.ndarray] = None,
```
Comment (patil-suraj) [inline src/transformers/models/t5/modeling_flax_t5.py:841]: `decoder_input_ids` shouldn't `None` for T5, we should raise if they are `None`

File: src/transformers/models/t5/modeling_flax_t5.py [hunk 10]
```diff
@@ -0,0 +1,1584 @@
...
+            decoder_attention_mask = jnp.ones_like(decoder_input_ids)
+
+        # Handle any PRNG if needed
+        rngs = {"dropout": dropout_rng} if dropout_rng is not None else {}
+
+        return self.module.apply(
+            {"params": params or self.params},
+            input_ids=jnp.array(input_ids, dtype="i4"),
+            attention_mask=jnp.array(attention_mask, dtype="i4"),
+            decoder_input_ids=jnp.array(decoder_input_ids, dtype="i4"),
```
Comment (patil-suraj) [inline src/transformers/models/t5/modeling_flax_t5.py:871]: This will crash if `decoder_input_ids` are `None`

File: src/transformers/models/t5/modeling_flax_t5.py [hunk 11]
```diff
@@ -0,0 +1,1584 @@
...
+
+
+def shift_tokens_right(input_ids: jnp.ndarray, pad_token_id: int, decoder_start_token_id: int) -> jnp.ndarray:
+    """
+    Shift input ids one token to the right.
+    """
+    shifted_input_ids = jnp.roll(input_ids, 1, axis=-1)
+    shifted_input_ids = jax.ops.index_update(shifted_input_ids, (..., 0), decoder_start_token_id)
+    # replace possible -100 values in labels by `pad_token_id`
+    shifted_input_ids = jnp.where(shifted_input_ids == -100, pad_token_id, shifted_input_ids)
```
Comment (patil-suraj) [inline src/transformers/models/t5/modeling_flax_t5.py:57]: actually this should not be necessary as we don't replace pad tokens with -100 in Flax

File: src/transformers/models/t5/modeling_flax_t5.py [hunk 12]
```diff
@@ -0,0 +1,1584 @@
...
+        # counter-act scaling in dot_product_attention_weights function
+        query_states *= jnp.sqrt(query_states.shape[-1])
+
+        # for fast decoding causal attention mask should be shifted
+        causal_attention_mask_shift = (
+            self.variables["cache"]["cache_index"] if (self.has_variable("cache", "cached_key") and self.causal) else 0
+        )
+        # create causal attention_mask; attention_mask has to be defined when model is causal
+        if self.causal:
+            causal_attention_mask = make_causal_mask(attention_mask, dtype="bool")
```
Comment (patil-suraj) [inline src/transformers/models/t5/modeling_flax_t5.py:316]: maybe add a check for `None` `attention_mask`

File: tests/test_modeling_flax_t5.py (hunk 6 above)
Comment (patil-suraj) [reply tests/test_modeling_flax_t5.py:461]: Nice!

File: tests/test_modeling_flax_t5.py (hunk 7 above)
Comment (patil-suraj) [reply tests/test_modeling_flax_t5.py:476]: Very cool!

**Review by patil-suraj — COMMENTED** [review]
Looks good to me! Left a few comments.

My only big suggestion is few modules are missing weight initializers,  think those are necessary for pre-training.

File: setup.py (hunk 1 above)
Comment (sgugger) [reply setup.py:237]: Fine by me too.

File: src/transformers/__init__.py [hunk 13]
```diff
@@ -1597,6 +1597,7 @@
             "FlaxRobertaPreTrainedModel",
         ]
     )
+    _import_structure["models.t5"].extend(["FlaxT5ForConditionalGeneration", "FlaxT5Model"])
```
Comment (sgugger) [inline src/transformers/__init__.py:1600]: Missing the PreTrainedModel, but I'm adding it in #12314

File: src/transformers/models/t5/modeling_flax_t5.py [hunk 14]
```diff
@@ -0,0 +1,1584 @@
...
+            # update key, value caches with our new 1d spatial slices
+            cur_index = cache_index.value
+            indices = (0,) * len(batch_dims) + (cur_index, 0, 0)
+            key = jax.lax.dynamic_update_slice(cached_key.value, key, indices)
+            value = jax.lax.dynamic_update_slice(cached_value.value, value, indices)
+            cached_key.value = key
+            cached_value.value = value
+            num_updated_cache_vectors = query.shape[1]
+            cache_index.value = cache_index.value + num_updated_cache_vectors
+            # causal mask for cached decoder self-attention: our single query position should only attend to those key positions that have already been generated and cached, not the remaining zero elements.
```
Comment (sgugger) [inline src/transformers/models/t5/modeling_flax_t5.py:249]: Nit: should go on two lines.

File: src/transformers/models/t5/modeling_flax_t5.py [hunk 15]
```diff
@@ -0,0 +1,1584 @@
...
+            attention_outputs = attention_outputs + cross_attention_outputs[1:]
+
+        # Apply Feed Forward layer
+        hidden_states = self.layer[-1](hidden_states, deterministic=deterministic)
+
+        outputs = (hidden_states,)
+
+        outputs = outputs + attention_outputs
+
+        return outputs  # hidden-states, present_key_value_states, (self-attention position bias), (self-attention weights), (cross-attention position bias), (cross-attention weights)
```
Comment (sgugger) [inline src/transformers/models/t5/modeling_flax_t5.py:530]: Nit: put the comment above and on several lines if needed.

File: src/transformers/models/t5/modeling_flax_t5.py [hunk 16]
```diff
@@ -0,0 +1,1584 @@
...
+
+            >>> tokenizer = T5Tokenizer.from_pretrained('t5-small')
+            >>> model = FlaxT5ForConditionalGeneration.from_pretrained('t5-small')
+
+            >>> input_ids = tokenizer('The <extra_id_0> walks in <extra_id_1> park', return_tensors='np').input_ids
+            >>> decoder_input_ids = tokenizer('<extra_id_0> cute dog <extra_id_1> the <extra_id_2> </s>', return_tensors='np').input_ids
+            >>> outputs = model(input_ids=input_ids, decoder_input_ids=decoder_input_ids)
+            >>> logits = outputs.logits
+
+            >>> input_ids = tokenizer("summarize: studies have shown that owning a dog is good for you ", return_tensors="np").input_ids  # Batch size 1
```
Comment (sgugger) [inline src/transformers/models/t5/modeling_flax_t5.py:1364]: Nit: Put the comment above for readability.

File: src/transformers/__init__.py (hunk 13 above)
Comment (patrickvonplaten) [reply src/transformers/__init__.py:1600]: ok not adding it here then

**Comment by caseybasichis** [comment]
Is a jax/flax byT5 planned?

Interested in both byT5 and jax... torn.

**Comment by patrickvonplaten** [comment]
You can already use ByT5 in jax/flax! Check the model card [here](https://huggingface.co/models?filter=jax&search=byt5)

**Comment by caseybasichis** [comment]
oh, my 🤗

**Comment by caseybasichis** [comment]
I'm having trouble finding the Jax model training and architecture definition.

Is this just loading a byT5 model into a regular T5 inference scaffolding?

My aim is to experiment with the training / masking code.

**Comment by patrickvonplaten** [comment]
Here is a test for FlaxByT5 that could help a bit: https://github.com/huggingface/transformers/blob/332a2458611751e7d9c4d7a21bc454299d50e160/tests/test_modeling_flax_t5.py#L432

Also we have `run_summarization` script in Flax that can be easily tweaked for any seq2seq task. And soon (Monday) we'll have FlaxT5 pretraining as well :-)
